File: src/python/src/flashalpha_quantconnect/exceptions.py
"""Exception hierarchy for flashalpha-quantconnect.

All bridge exceptions inherit from FlashAlphaQuantConnectException so users
can catch the family in one block. Each exception has a fixed error code
that maps to docs/troubleshooting.md.
"""


class FlashAlphaQuantConnectException(Exception):
    """Base for all bridge exceptions."""

    error_code: str = "FA-000"

    @property
    def doc_url(self) -> str:
        return (
            "https://github.com/FlashAlpha-lab/flashalpha-quantconnect/"
            f"blob/main/docs/troubleshooting.md#{self.error_code.lower()}"
        )


class FlashAlphaAuthMissingException(FlashAlphaQuantConnectException):
    error_code = "FA-AUTH-001"

    def __init__(self) -> None:
        super().__init__(
            "FlashAlpha API key not found. Set FLASHALPHA_API_KEY env var, "
            "flashalpha_quantconnect.config.api_key, or QC parameter "
            "'flashalpha-api-key'.",
        )


class FlashAlphaUnauthorizedException(FlashAlphaQuantConnectException):
    error_code = "FA-AUTH-002"

    def __init__(self, endpoint: str, key_tail4: str) -> None:
        super().__init__(
            f"FlashAlpha rejected the API key (…{key_tail4}) on endpoint {endpoint}.",
        )


class FlashAlphaRateLimitedException(FlashAlphaQuantConnectException):
    error_code = "FA-RATE-001"

    def __init__(self, endpoint: str) -> None:
        super().__init__(
            f"FlashAlpha rate-limited the request to {endpoint} after retries exhausted.",
        )


class FlashAlphaNetworkException(FlashAlphaQuantConnectException):
    error_code = "FA-NET-001"

    def __init__(self, endpoint: str) -> None:
        super().__init__(f"Network error talking to FlashAlpha at {endpoint}.")

File: src/csharp/FlashAlpha.QuantConnect/Exceptions.cs
using System;

namespace FlashAlpha.QuantConnect;

public class FlashAlphaQuantConnectException : Exception
{
    public string ErrorCode { get; }
    public string DocUrl => $"https://github.com/FlashAlpha-lab/flashalpha-quantconnect/blob/main/docs/troubleshooting.md#{ErrorCode.ToLowerInvariant()}";

    protected FlashAlphaQuantConnectException(string errorCode, string message, Exception? inner = null)
        : base(message, inner)
    {
        ErrorCode = errorCode;
    }
}

public sealed class FlashAlphaAuthMissingException : FlashAlphaQuantConnectException
{
    public FlashAlphaAuthMissingException()
        : base("FA-AUTH-001", "FlashAlpha API key not found. Set FLASHALPHA_API_KEY env var, FlashAlphaConfig.ApiKey, or QC parameter 'flashalpha-api-key'.") { }
}

public sealed class FlashAlphaUnauthorizedException : FlashAlphaQuantConnectException
{
    public FlashAlphaUnauthorizedException(string endpoint, string keyTail4)
        : base("FA-AUTH-002", $"FlashAlpha rejected the API key (…{keyTail4}) on endpoint {endpoint}. See troubleshooting docs.") { }
}

public sealed class FlashAlphaRateLimitedException : FlashAlphaQuantConnectException
{
    public FlashAlphaRateLimitedException(string endpoint, Exception inner)
        : base("FA-RATE-001", $"FlashAlpha rate-limited the request to {endpoint} after retries exhausted.", inner) { }
}

public sealed class FlashAlphaNetworkException : FlashAlphaQuantConnectException
{
    public FlashAlphaNetworkException(string endpoint, Exception inner)
        : base("FA-NET-001", $"Network error talking to FlashAlpha at {endpoint}.", inner) { }
}
